### PR TITLE
POL-261 Trigger lifecycle

### PR DESCRIPTION
**Pull request by burmanm**

### CODE REVIEW

**Comment by burmanm** [comment]
@pilhuhn This is the current output. Would it be enough for the"last triggered" timestamp?

```json
{
  "trigger": {
    "tenantId": "tutorial",
    "id": "detect-floating-4-tutorial4",
    "name": "Node with no infra third",
    "description": "These hosts are not allocated to any known infrastructure provider",
    "type": "STANDARD",
    "eventType": "ALERT",
    "eventCategory": null,
    "eventText": null,
    "severity": "MEDIUM",
    "actions": [
      {
        "tenantId": "tutorial",
        "actionPlugin": "hooks",
        "actionId": "_managed-instance-hooks-19c9899847a8bf14"
      }
    ],
    "autoDisable": false,
    "autoEnable": false,
    "autoResolve": false,
    "autoResolveAlerts": true,
    "autoResolveMatch": "ALL",
    "enabled": true,
    "firingMatch": "ALL",
    "source": "_none_",
    "lifecycle": [
      {
        "status": "CREATED",
        "stime": 1588926428630
      },
      {
        "status": "ALERT_GENERATE",
        "stime": 1588926432637
      }
    ]
  },
  "conditions": [
    {
      "tenantId": "tutorial",
      "triggerId": "detect-floating-4-tutorial4",
      "triggerMode": "FIRING",
      "type": "EVENT",
      "conditionSetSize": 1,
      "conditionSetIndex": 1,
      "conditionId": "tutorial-detect-floating-4-tutorial4-FIRING-1-1",
      "displayString": "platform.inventory.host-egress matches [null]",
      "lastEvaluation": 1588926432534,
      "dataId": "platform.inventory.host-egress",
      "expression": null
    }
  ]
}
```

**Comment by pilhuhn** [comment]
I guess this is a _lot_ more than what I need :)

Querying would still work by passing individual ids as queryParam?

**Comment by burmanm** [comment]
Yes, this is returned as part of normal trigger queries. The above one was with ``curl -H 'Hawkular-Tenant: tutorial'  -X GET http://localhost:8084/hawkular/alerts/triggers/trigger/detect-floating-4-tutorial4 | jq .``

**Comment by burmanm** [comment]
And there's of course no need to get full trigger, a trigger definition is enough:

```
[michael@nina alerts-cleanup]$ curl -s -H 'Hawkular-Tenant: tutorial'  -X GET http://localhost:8084/hawkular/alerts/triggers/detect-floating-4-tutorial4 | jq .
{
  "tenantId": "tutorial",
  "id": "detect-floating-4-tutorial4",
  "name": "Node with no infra third",
  "description": "These hosts are not allocated to any known infrastructure provider",
  "type": "STANDARD",
  "eventType": "ALERT",
  "eventCategory": null,
  "eventText": null,
  "severity": "MEDIUM",
  "actions": [
    {
      "tenantId": "tutorial",
      "actionPlugin": "hooks",
      "actionId": "_managed-instance-hooks-19c9899847a8bf14"
    }
  ],
  "autoDisable": false,
  "autoEnable": false,
  "autoResolve": false,
  "autoResolveAlerts": true,
  "autoResolveMatch": "ALL",
  "enabled": true,
  "firingMatch": "ALL",
  "source": "_none_",
  "lifecycle": [
    {
      "status": "CREATED",
      "stime": 1588926428630
    },
    {
      "status": "ALERT_GENERATE",
      "stime": 1588926432637
    }
  ]
}
[michael@nina alerts-cleanup]$
```

**Comment by pilhuhn** [comment]
How would I query 5 triggers at once? The UI-backend needs to display pages of 50 triggers, so having a single call per trigger is not optimal.

**Comment by burmanm** [comment]
Using triggerIds: https://github.com/RedHatInsights/policies-engine/blob/0420840039595f0cea786feca483234472e40143/external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java#L640

**Comment by burmanm** [comment]
Or obviously, just get all the triggers for tenant and use the paging parameters to get a sub slice.

**Comment by pilhuhn** [comment]
I missed that this was already merged last week :-/

File: api/src/main/java/org/hawkular/alerts/api/model/trigger/Trigger.java [hunk 1]
```diff
@@ -630,4 +665,7 @@ public String toString() {
                 + ", mode=" + mode + ", tags=" + tags + "]";
     }
 
+    public enum TriggerLifecycle {
+        ALERT_GENERATE, AUTO_DISABLE, AUTO_RESOLVE, DISABLE, ENABLE, CREATED, MODIFIED
```
Comment (pilhuhn) [inline api/src/main/java/org/hawkular/alerts/api/model/trigger/Trigger.java:669]: Should those all in present or all in past?

ENABLE vs createD & modifieD

**Comment by pilhuhn** [comment]
This is again creating a possibly unbounded list of lifecycle events. Is there a way to limit what is returned?

**Comment by burmanm** [comment]
No, there isn't. It's a single document at this point.

File: api/src/main/java/org/hawkular/alerts/api/model/trigger/Trigger.java (hunk 1 above)
Comment (burmanm) [reply api/src/main/java/org/hawkular/alerts/api/model/trigger/Trigger.java:669]: Yeah, should be in the present form (as they happen).

**Pull request merged**